File: pkg/ikuai_api4/stream_domain.go
package ikuai_api4

import (
	"encoding/json"
	"errors"
	"ikuai-bypass/pkg/ikuai_common"
	"strconv"
	"strings"
)

const FuncNameStreamDomain = "stream_domain"

// 4.0 域名分流专用结构
// Specific structures for 4.0 stream_domain
type streamDomain4 struct {
	ID        int    `json:"id"`
	Enabled   string `json:"enabled"`
	Tagname   string `json:"tagname"`
	Interface string `json:"interface"`
	Comment   string `json:"comment"`
	Prio      int    `json:"prio"`
	SrcAddr   struct {
		Custom interface{} `json:"custom"`
		Object interface{} `json:"object"`
	} `json:"src_addr"`
	Domain struct {
		Custom interface{} `json:"custom"`
		Object interface{} `json:"object"`
	} `json:"domain"`
	Time struct {
		Custom []struct {
			Type      string `json:"type"`
			Weekdays  string `json:"weekdays"`
			StartTime string `json:"start_time"`
			EndTime   string `json:"end_time"`
			Comment   string `json:"comment"`
		} `json:"custom"`
		Object interface{} `json:"object"`
	} `json:"time"`
}

func (i *IKuai) AddStreamDomain(iface, tag, srcAddr, srcAddrOptIpGroup, domains string, index int) error {
	domains = strings.TrimSpace(domains)
	domainList := strings.Split(domains, ",")

	var srcCustom []string
	var srcObjects []ipGroupObject4

	if strings.TrimSpace(srcAddrOptIpGroup) != "" {
		names := strings.Split(srcAddrOptIpGroup, ",")
		var resolvedNames []string
		for _, name := range names {
			name = strings.TrimSpace(name)
			if name == "" {
				continue
			}
			matches, err := i.GetAllIKuaiBypassIpGroupNamesByName(name)
			if err == nil {
				resolvedNames = append(resolvedNames, matches...)
			}
		}
		srcObjects = i.resolveIpGroupObjects(resolvedNames)
		srcCustom = []string{}
	} else {
		srcAddr = strings.TrimSpace(srcAddr)
		var srcAddrList []string
		if srcAddr != "" {
			srcAddrList = strings.Split(srcAddr, ",")
		}
		var srcObjectNames []string
		srcCustom, srcObjectNames = CategorizeAddrs(srcAddrList)
		srcObjects = i.resolveIpGroupObjects(srcObjectNames)
	}

	uniqueTagname := buildIndexedTagName(tag, index)

	param := map[string]interface{}{
		"enabled":   "yes",
		"tagname":   uniqueTagname,
		"interface": iface,
		"src_addr": map[string]interface{}{
			"custom": srcCustom,
			"object": srcObjects,
		},
		"domain": map[string]interface{}{
			"custom": domainList,
			"object": []interface{}{},
		},
		"comment": ikuai_common.NEW_COMMENT,
		"time": map[string]interface{}{
			"custom": []map[string]interface{}{
				{
					"type":       "weekly",
					"weekdays":   "1234567",
					"start_time": "00:00",
					"end_time":   "23:59",
					"comment":    "",
				},
			},
			"object": []interface{}{},
		},
		"prio": 31,
	}

	req := CallReq{
		FuncName: FuncNameStreamDomain,
		Action:   "add",
		Param:    &param,
	}

	resp := CallResp{}
	err := postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return err
	}
	if resp.Code != 0 {
		return errors.New(resp.Message)
	}
	return nil
}

func (i *IKuai) EditStreamDomain(iface, tag, srcAddr, srcAddrOptIpGroup, domains string, index int, id int) error {
	domains = strings.TrimSpace(domains)
	domainList := strings.Split(domains, ",")

	var srcCustom []string
	var srcObjects []ipGroupObject4
	if strings.TrimSpace(srcAddrOptIpGroup) != "" {
		names := strings.Split(srcAddrOptIpGroup, ",")
		var resolvedNames []string
		for _, name := range names {
			name = strings.TrimSpace(name)
			if name == "" {
				continue
			}
			matches, err := i.GetAllIKuaiBypassIpGroupNamesByName(name)
			if err == nil {
				resolvedNames = append(resolvedNames, matches...)
			}
		}
		srcObjects = i.resolveIpGroupObjects(resolvedNames)
		srcCustom = []string{}
	} else {
		srcAddr = strings.TrimSpace(srcAddr)
		var srcAddrList []string
		if srcAddr != "" {
			srcAddrList = strings.Split(srcAddr, ",")
		}
		var srcObjectNames []string
		srcCustom, srcObjectNames = CategorizeAddrs(srcAddrList)
		srcObjects = i.resolveIpGroupObjects(srcObjectNames)
	}

	uniqueTagname := buildIndexedTagName(tag, index)

	param := map[string]interface{}{
		"enabled":   "yes",
		"tagname":   uniqueTagname,
		"interface": iface,
		"src_addr": map[string]interface{}{
			"custom": srcCustom,
			"object": srcObjects,
		},
		"domain": map[string]interface{}{
			"custom": domainList,
			"object": []interface{}{},
		},
		"comment": ikuai_common.NEW_COMMENT,
		"time": map[string]interface{}{
			"custom": []map[string]interface{}{
				{
					"type":       "weekly",
					"weekdays":   "1234567",
					"start_time": "00:00",
					"end_time":   "23:59",
					"comment":    "",
				},
			},
			"object": []interface{}{},
		},
		"prio": 31,
		"id":   id,
	}

	req := CallReq{
		FuncName: FuncNameStreamDomain,
		Action:   "edit",
		Param:    &param,
	}

	resp := CallResp{}
	err := postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return err
	}
	if resp.Code != 0 {
		return errors.New(resp.Message)
	}
	return nil
}

func (i *IKuai) GetStreamDomainMap(tag string) (result map[int]int, err error) {
	result = make(map[int]int)
	var data []ikuai_common.StreamDomainData
	data, err = i.ShowStreamDomainByTagName("")
	if err != nil {
		return nil, err
	}

	baseName := buildTagName(tag)
	for _, d := range data {
		if matchTagNameFilter(tag, d.TagName, d.Comment) {
			suffix := strings.TrimPrefix(d.TagName, baseName)
			if idx, err := strconv.Atoi(suffix); err == nil {
				result[idx] = d.ID
			}
		}
	}
	return result, nil
}

func (i *IKuai) ShowStreamDomainByTagName(tagName string) (result []ikuai_common.StreamDomainData, err error) {
	param := map[string]interface{}{
		"TYPE":  "total,data",
		"limit": "0,1000",
	}
	req := CallReq{
		FuncName: FuncNameStreamDomain,
		Action:   "show",
		Param:    &param,
	}

	var data4 []streamDomain4
	resp := CallResp{Results: &CallRespData{Data: &data4}}
	err = postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return
	}
	if resp.Code != 0 {
		err = errors.New(resp.Message)
		return
	}

	// 将 4.0 结构转换为通用结构
	// Convert 4.0 structure to common structure
	for _, d := range data4 {
		if matchTagNameFilter(tagName, d.Tagname, d.Comment) {
			srcs := append(toStringList(d.SrcAddr.Custom), toStringList(d.SrcAddr.Object)...)
			domains := append(toStringList(d.Domain.Custom), toStringList(d.Domain.Object)...)

			item := ikuai_common.StreamDomainData{
				ID:        d.ID,
				Enabled:   d.Enabled,
				Comment:   d.Comment,
				TagName:   d.Tagname,
				Interface: d.Interface,
				Domain:    strings.Join(domains, ","),
				SrcAddr:   strings.Join(srcs, ","),
			}
			// 尝试还原 Week 和 Time 字段
			if len(d.Time.Custom) > 0 {
				item.Week = d.Time.Custom[0].Weekdays
				item.Time = d.Time.Custom[0].StartTime + "-" + d.Time.Custom[0].EndTime
			}
			result = append(result, item)
		}
	}

	return
}

func (i *IKuai) DelStreamDomain(id string) error {
	param := struct {
		Id string `json:"id"`
	}{
		Id: id,
	}
	req := CallReq{
		FuncName: FuncNameStreamDomain,
		Action:   "del",
		Param:    &param,
	}
	resp := CallResp{}
	err := postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return err
	}
	if resp.Code != 0 {
		// DEBUG: 打印完整错误响应
		respJson, _ := json.Marshal(resp)
		i.L.Error("DEBUG:错误响应", "code=%d, message=%s, full_response=%s", resp.Code, resp.Message, string(respJson))
		return errors.New(resp.Message)
	}
	return nil
}

// GetStreamDomainAll 批量查询并返回逗号分隔的 ID
// Batch query and return comma-separated IDs
func (i *IKuai) GetStreamDomainAll(tag string) (preIds string, err error) {
	i.L.Info("QUERY:查询列表", "Querying domain streaming rules (Prefix: %s, Tag: %s)", ikuai_common.NAME_PREFIX_IKB, tag)
	preIds = ""
	err = nil
	var data []ikuai_common.StreamDomainData
	data, err = i.ShowStreamDomainByTagName("")
	if err != nil {
		return
	}
	var ids []string
	for _, d := range data {
		if matchTagNameFilter(tag, d.TagName, d.Comment) {
			ids = append(ids, strconv.Itoa(d.ID))
		}
	}
	if len(ids) <= 0 {
		return
	}
	id := strings.Join(ids, ",")
	preIds = "||" + id
	return
}

// DelStreamDomainFromPreIds 从预备删除的id中删除
func (i *IKuai) DelStreamDomainFromPreIds(preIds string) (err error) {
	arr := strings.Split(preIds, "||")
	for _, id := range arr {
		if len(id) < 1 {
			continue
		}
		err = i.DelStreamDomain(id)
		if err != nil {
			return
		}
	}
	return

}

// DelStreamDomainAll 删除所有的域名分流规则
func (i *IKuai) DelStreamDomainAll(cleanTag string) (err error) {
	for {
		var data []ikuai_common.StreamDomainData
		data, err = i.ShowStreamDomainByTagName("")
		if err != nil {
			return
		}
		var ids []string
		for _, d := range data {
			if matchCleanTag(cleanTag, d.Comment, d.TagName) {
				ids = append(ids, strconv.Itoa(d.ID))
			}
		}
		if len(ids) <= 0 {
			return
		}
		id := strings.Join(ids, ",")
		err = i.DelStreamDomain(id)
		if err != nil {
			return
		}
	}
}

File: rust/core/src/ikuai/stream_domain.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::clean::match_clean_tag;
use super::ip_group;
use super::tag_name::{build_indexed_tag_name, match_tag_name_filter};
use super::types::{IKuaiClient, IKuaiError, StreamDomainData, FUNC_NAME_STREAM_DOMAIN, NEW_COMMENT};
use super::utils::{categorize_addrs, to_string_list};

#[derive(Debug, Deserialize)]
struct AddrBlock {
    custom: Value,
    object: Value,
}

#[derive(Debug, Deserialize)]
struct TimeCustom {
    weekdays: String,
    start_time: String,
    end_time: String,
}

#[derive(Debug, Deserialize)]
struct TimeBlock {
    custom: Vec<TimeCustom>,
    object: Value,
}

#[derive(Debug, Deserialize)]
struct StreamDomain4 {
    id: i64,
    enabled: String,
    #[serde(rename = "tagname")]
    tagname: String,
    interface: String,
    comment: String,
    prio: i64,
    #[serde(rename = "src_addr")]
    src_addr: AddrBlock,
    domain: AddrBlock,
    time: TimeBlock,
}

#[derive(Debug, Serialize)]
struct ShowParam {
    #[serde(rename = "TYPE")]
    r#type: String,
    limit: String,
}

#[derive(Debug, Serialize)]
struct DelParam {
    id: String,
}

pub async fn show_stream_domain_by_tag_name(
    api: &IKuaiClient,
    tag_name: &str,
) -> Result<Vec<StreamDomainData>, IKuaiError> {
    let param = ShowParam {
        r#type: "total,data".to_string(),
        limit: "0,1000".to_string(),
    };
    let resp = api
        .call::<_, Vec<StreamDomain4>>(FUNC_NAME_STREAM_DOMAIN, "show", &param)
        .await?;
    let data = resp.results.ok_or(IKuaiError::InvalidResponse)?.data;
    let mut out = Vec::new();
    for d in data {
        if !match_tag_name_filter(tag_name, &d.tagname, &d.comment) {
            continue;
        }
        let srcs = {
            let mut v = to_string_list(&d.src_addr.custom);
            v.extend(to_string_list(&d.src_addr.object));
            v
        };
        let domains = {
            let mut v = to_string_list(&d.domain.custom);
            v.extend(to_string_list(&d.domain.object));
            v
        };
        let mut item = StreamDomainData {
            id: d.id,
            enabled: d.enabled,
            comment: d.comment,
            tag_name: d.tagname,
            interface: d.interface,
            domain: domains.join(","),
            src_addr: srcs.join(","),
            week: String::new(),
            time: String::new(),
        };
        if let Some(t) = d.time.custom.first() {
            item.week = t.weekdays.clone();
            item.time = format!("{}-{}", t.start_time, t.end_time);
        }
        out.push(item);
    }
    Ok(out)
}

pub async fn add_stream_domain(
    api: &IKuaiClient,
    iface: &str,
    tag: &str,
    src_addr: &str,
    src_addr_opt_ipgroup: &str,
    domains: &str,
    index: i64,
) -> Result<(), IKuaiError> {
    let domain_list: Vec<String> = domains
        .trim()
        .split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    let (src_custom, src_objects) = resolve_src_addrs(api, src_addr, src_addr_opt_ipgroup).await?;
    let unique_tagname = build_indexed_tag_name(tag, index);

    let param = serde_json::json!({
        "enabled": "yes",
        "tagname": unique_tagname,
        "interface": iface,
        "src_addr": {"custom": src_custom, "object": src_objects},
        "domain": {"custom": domain_list, "object": []},
        "comment": NEW_COMMENT,
        "time": {
            "custom": [{"type":"weekly","weekdays":"1234567","start_time":"00:00","end_time":"23:59","comment":""}],
            "object": []
        },
        "prio": 31
    });
    let _ = api
        .call::<_, serde_json::Value>(FUNC_NAME_STREAM_DOMAIN, "add", &param)
        .await?;
    Ok(())
}

pub async fn edit_stream_domain(
    api: &IKuaiClient,
    iface: &str,
    tag: &str,
    src_addr: &str,
    src_addr_opt_ipgroup: &str,
    domains: &str,
    index: i64,
    id: i64,
) -> Result<(), IKuaiError> {
    let domain_list: Vec<String> = domains
        .trim()
        .split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    let (src_custom, src_objects) = resolve_src_addrs(api, src_addr, src_addr_opt_ipgroup).await?;
    let unique_tagname = build_indexed_tag_name(tag, index);

    let param = serde_json::json!({
        "enabled": "yes",
        "tagname": unique_tagname,
        "interface": iface,
        "src_addr": {"custom": src_custom, "object": src_objects},
        "domain": {"custom": domain_list, "object": []},
        "comment": NEW_COMMENT,
        "time": {
            "custom": [{"type":"weekly","weekdays":"1234567","start_time":"00:00","end_time":"23:59","comment":""}],
            "object": []
        },
        "prio": 31,
        "id": id
    });
    let _ = api
        .call::<_, serde_json::Value>(FUNC_NAME_STREAM_DOMAIN, "edit", &param)
        .await?;
    Ok(())
}

pub async fn del_stream_domain(api: &IKuaiClient, id_csv: &str) -> Result<(), IKuaiError> {
    let param = DelParam {
        id: id_csv.to_string(),
    };
    let _ = api
        .call::<_, serde_json::Value>(FUNC_NAME_STREAM_DOMAIN, "del", &param)
        .await?;
    Ok(())
}

pub async fn get_stream_domain_map(
    api: &IKuaiClient,
    tag: &str,
) -> Result<std::collections::HashMap<i64, i64>, IKuaiError> {
    let data = show_stream_domain_by_tag_name(api, "").await?;
    let base = super::tag_name::build_tag_name(tag);
    let mut out = std::collections::HashMap::new();
    for d in data {
        if !match_tag_name_filter(tag, &d.tag_name, &d.comment) {
            continue;
        }
        let suffix = d.tag_name.trim().trim_start_matches(&base);
        if let Ok(idx) = suffix.parse::<i64>() {
            out.insert(idx, d.id);
        }
    }
    Ok(out)
}

pub async fn del_stream_domain_all(api: &IKuaiClient, clean_tag: &str) -> Result<(), IKuaiError> {
    loop {
        let data = show_stream_domain_by_tag_name(api, "").await?;
        let ids: Vec<String> = data
            .into_iter()
            .filter(|d| match_clean_tag(clean_tag, &d.comment, &d.tag_name))
            .map(|d| d.id.to_string())
            .collect();
        if ids.is_empty() {
            return Ok(());
        }
        del_stream_domain(api, &ids.join(",")).await?;
    }
}

async fn resolve_src_addrs(
    api: &IKuaiClient,
    src_addr: &str,
    src_addr_opt_ipgroup: &str,
) -> Result<(Vec<String>, Vec<Value>), IKuaiError> {
    if !src_addr_opt_ipgroup.trim().is_empty() {
        let mut resolved = Vec::new();
        for name in src_addr_opt_ipgroup.split(',') {
            let name = name.trim();
            if name.is_empty() {
                continue;
            }
            if let Ok(matches) = ip_group::get_all_ikuai_bypass_ip_group_names_by_name(api, name).await {
                resolved.extend(matches);
            }
        }
        let objects = resolve_ip_group_objects(api, &resolved).await?;
        return Ok((Vec::new(), objects));
    }

    let mut src_list = Vec::new();
    let src_addr = src_addr.trim();
    if !src_addr.is_empty() {
        src_list.extend(src_addr.split(',').map(|s| s.trim().to_string()));
    }
    let (custom, obj_names) = categorize_addrs(&src_list);
    let objects = resolve_ip_group_objects(api, &obj_names).await?;
    Ok((custom, objects))
}

async fn resolve_ip_group_objects(api: &IKuaiClient, names: &[String]) -> Result<Vec<Value>, IKuaiError> {
    if names.is_empty() {
        return Ok(Vec::new());
    }
    let groups = ip_group::show_ip_group_by_tag_name(api, "").await?;
    let mut map = std::collections::HashMap::new();
    for g in groups {
        map.insert(g.group_name.clone(), g.id);
    }
    let mut out = Vec::new();
    for name in names {
        if let Some(id) = map.get(name) {
            out.push(serde_json::json!({
                "type": 0,
                "gid": format!("IPGP{}", id),
                "gp_name": name
            }));
        }
    }
    Ok(out)
}
